src/neps/optimizers/bayesian_optimization/acquisition_samplers/freeze_thaw_sampler.py

```diff
@@ -67,31 +91,63 @@ def _sample_new(
             new_configs, index=range(index_from, index_from + len(new_configs))
         )
 
-    def sample(self, acquisition_function=None, n: int = None) -> pd.Series:
+    def sample(
+        self,
+        acquisition_function=None,
+        n: int = None,
+        set_new_sample_fidelity: int | float = None,
+    ) -> list():
+        """Samples a new set and returns the total set of observed + new configs."""
         partial_configs = self.observations.get_partial_configs_at_max_seen()
+        new_configs = self._sample_new(
+            index_from=self.observations.next_config_id(), n=n, ignore_fidelity=False
+        )
 
-        new_configs = self._sample_new(index_from=self.observations.next_config_id(), n=n)
-
-        # Set fidelity for observed configs
+        def __sample_single_new_tabular(index: int):
+            """
+            A function to use in a list comprehension to slightly speed up
+            the sampling process when self.SAMPLE_TO_DRAW is large
+            """
+            config = self.pipeline_space.sample(
+                patience=self.patience, user_priors=False, ignore_fidelity=False
+            )
+            config["id"].value = _new_configs[index]
+            config.fidelity.value = set_new_sample_fidelity
+            return config
+
+        if self.is_tabular:
+            _n = n if n is not None else self.SAMPLES_TO_DRAW
+            _partial_ids = {conf["id"].value for conf in partial_configs}
+            _all_ids = set(self.pipeline_space.custom_grid_table.index.values)
+            # accounting for unseen configs only
+            _n = min(_n, len(_all_ids - _partial_ids))
+            _new_configs = np.random.choice(
+                list(_all_ids - _partial_ids), size=_n, replace=False
+            )
+            new_configs = [__sample_single_new_tabular(i) for i in range(_n)]
```

Comment:
Is there a way to get the entire tabular grid once? For me, this line can return duplicates in the sampled configurations.

Reply:
As far as my understanding goes, this should never return duplicates. Do you know if it returns multiple identical configurations in the `new_configs` or duplicate configs with already explored ones?

Reply:
I was wrong, sorry. It doesn't return duplicates indeed.